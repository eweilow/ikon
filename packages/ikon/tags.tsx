import { Sema } from "async-sema";
import { existsSync, promises } from "fs";
import hasha from "hasha";
import mkdirp from "mkdirp";
import { cpus } from "os";
import { join } from "path";
import puppeteer, { Browser, Page } from "puppeteer";

import { generateFavicon } from "./favicon";
import { getIcons, renderIcon } from "./render";
import { DefaultAppIconName, DefaultLaunchScreenName } from "./sizes";
import { IconGenerationComponent } from "./types";

export async function generateTags(
  Component: IconGenerationComponent,
  publicPath: string,
  outDir: string,
  maxConcurrency: number,
  onTagCreated?: (tag: string) => void,
  onImageCreated?: (name: string, buffer: Buffer, publicName: string) => void,
  skipBuildIfPossible: boolean = false
) {
  mkdirp.sync(outDir);

  const browsersCount = Math.min(maxConcurrency, Math.max(1, (cpus() || []).length));
  const pagesPerBrowser = 1;

  const sema = new Sema(browsersCount * pagesPerBrowser);
  const pages: Page[] = [];
  const browsers: Browser[] = [];

  let booted: boolean = false;
  let bootPromise!: Promise<void>;
  async function bootPuppeteerIfNeeded() {
    if (booted) {
      return bootPromise;
    }
    async function innerBoot() {
      console.log("Booting %d browsers", browsersCount);

      for (let i = 0; i < browsersCount; i++) {
        const browser = await puppeteer.launch({
          executablePath: process.env.PUPPETEER_EXECUTABLE_PATH,
          timeout: 5000,
          args: [
            "--no-sandbox",
            "--disable-setuid-sandbox",
            "--disable-background-timer-throttling",
            "--disable-backgrounding-occluded-windows",
            "--disable-renderer-backgrounding",
            "--disable-dev-shm-usage",
          ],
          pipe: true,
        });
        browsers.push(browser);

        for (let p = 0; p < pagesPerBrowser; p++) {
          const page = await browser.newPage();
          pages.push(page);
        }
      }
      console.log("Booted %d pages", pages.length);
    }
    bootPromise = innerBoot();
    booted = true;
    await bootPromise;
  }

  const tags: string[] = [];
  function tagDidComplete(tag: string) {
    if (onTagCreated != null) {
      onTagCreated(tag);
    }
    tags.push(tag);
  }

  function imageDidComplete(name: string, buffer: Buffer, publicName: string) {
    if (onImageCreated != null) {
      onImageCreated(name, buffer, publicName);
    }
  }

  try {
    const icons = getIcons();
    const filePaths = await Promise.all(
      icons.map(async (icon) => {
        await sema.acquire();
        try {
          const fileName = join(outDir, icon.id + ".png");

          let src!: Buffer;
          if (skipBuildIfPossible && existsSync(fileName)) {
            src = await promises.readFile(fileName);
          } else {
            await bootPuppeteerIfNeeded();
            const page = pages.shift() as Page;
            try {
              src = await renderIcon(icon, Component, page);
            } finally {
              pages.push(page);
            }
          }

          let hash: string = "";
          if (process.env.NODE_ENV !== "test") {
            hash =
              "?h=" +
              hasha(src, {
                algorithm: "sha1",
                encoding: "hex",
              }).slice(0, 8);
          } else {
            hash = `?h=test`;
          }

          const publicName = `${publicPath}/${icon.id}.png${hash}`;
          await promises.writeFile(fileName, src);
          imageDidComplete(fileName, src, publicName);

          if (icon.type === "splash") {
            tagDidComplete(
              `<link rel="apple-touch-icon" media="(device-width: ${
                icon.width
              }) and (device-height: ${icon.height}) and (orientation: ${
                icon.width < icon.height ? "portrait" : "landscape"
              }) and (-webkit-device-pixel-ratio: ${icon.pixelRatio})" href="${publicName}">`
            );
            if (icon.id === DefaultLaunchScreenName) {
              tagDidComplete(`<link rel="apple-touch-icon" href="${publicName}">`);
            }
          } else {
            if (icon.name.includes("apple")) {
              tagDidComplete(
                `<link rel="apple-touch-icon" sizes="${icon.width * icon.pixelRatio}x${
                  icon.height * icon.pixelRatio
                }" href="${publicName}">`
              );
              if (icon.id === DefaultAppIconName) {
                tagDidComplete(`<link rel="apple-touch-icon" href="${publicName}">`);
              }
            } else {
              tagDidComplete(
                `<link rel="icon" type="image/png" sizes="${icon.width * icon.pixelRatio}x${
                  icon.height * icon.pixelRatio
                }" href="${publicName}">`
              );
            }
          }
          return fileName;
        } finally {
          sema.release();
        }
      })
    );

    const favicon = await generateFavicon(icons, filePaths);
    const faviconName = join(outDir, "favicon.ico");
    await promises.writeFile(faviconName, favicon);

    let faviconHash: string = "";
    if (process.env.NODE_ENV !== "test") {
      faviconHash =
        "?h=" +
        hasha(favicon, {
          algorithm: "sha1",
          encoding: "hex",
        }).slice(0, 8);
    } else {
      faviconHash = `?h=test`;
    }

    tagDidComplete(`<link rel="shortcut icon" href="${publicPath}/favicon.ico${faviconHash}">`);
    imageDidComplete(faviconName, favicon, `${publicPath}/favicon.ico${faviconHash}`);
  } finally {
    for (const browser of browsers) {
      await browser.close();
    }
  }
}
