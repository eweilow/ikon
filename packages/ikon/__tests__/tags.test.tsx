import { existsSync } from "fs";
import { toMatchImageSnapshot } from "jest-image-snapshot";
import { basename, join } from "path";
import React from "react";
import { DirectoryResult, dir } from "tmp-promise";

import { generateTags } from "../tags";

expect.extend({ toMatchImageSnapshot });

describe("generateTags", () => {
  let tmpDir: DirectoryResult;
  jest.setTimeout(60000);

  beforeAll(async () => {
    tmpDir = await dir({ unsafeCleanup: true });
  });

  afterEach(async () => {
    await tmpDir.cleanup();
  });

  it("renders correctly", async () => {
    const tags: string[] = [];
    const images: { name: string; buffer: Buffer; publicName: string }[] = [];

    await generateTags(
      (props) => {
        const r = Math.floor(props.width % 255);
        const g = Math.floor(props.height % 255);
        const b = Math.floor((props.pixelRatio * 100) % 255);
        return (
          <div
            style={{
              height: "100%",
              width: "100%",
              color: "#" + [r, g, b].map((s) => s.toString(16)).join(""),
              background: "#" + [r, g, b].map((s) => (255 - s).toString(16)).join(""),
              textAlign: "center",
              position: "relative",
            }}
          >
            <div
              style={{
                position: "absolute",
                top: "50%",
                left: "50%",
                marginLeft: -8,
                marginTop: -8,
                height: 16,
                width: 16,
                borderRadius: 8,
                background: "white",
              }}
            />
          </div>
        );
      },
      "/",
      tmpDir.path,
      1,
      (tag) => tags.push(tag),
      (name, buffer, publicName) => images.push({ name, buffer, publicName }),
      false
    );

    for (const image of images) {
      const fileName = basename(image.name);
      if (fileName.endsWith(".ico")) {
        continue;
      }
      expect(image.buffer).toMatchImageSnapshot({
        customSnapshotIdentifier: fileName,
        blur: 1,
        failureThreshold: 0.01,
        failureThresholdType: "percent",
      });

      expect(existsSync(join(tmpDir.path, fileName))).toBe(true);
    }

    expect(tags).toMatchSnapshot("tags");
  });
});
