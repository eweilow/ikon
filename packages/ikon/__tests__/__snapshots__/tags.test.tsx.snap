// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateTags renders correctly: tags 1`] = `
[
  "<link rel="icon" type="image/png" sizes="16x16" href="//favicon-16-16.png?h=test">",
  "<link rel="icon" type="image/png" sizes="24x24" href="//favicon-24-24.png?h=test">",
  "<link rel="icon" type="image/png" sizes="32x32" href="//favicon-32-32.png?h=test">",
  "<link rel="icon" type="image/png" sizes="48x48" href="//favicon-48-48.png?h=test">",
  "<link rel="icon" type="image/png" sizes="64x64" href="//favicon-64-64.png?h=test">",
  "<link rel="icon" type="image/png" sizes="72x72" href="//favicon-72-72.png?h=test">",
  "<link rel="icon" type="image/png" sizes="96x96" href="//favicon-96-96.png?h=test">",
  "<link rel="icon" type="image/png" sizes="128x128" href="//favicon-128-128.png?h=test">",
  "<link rel="icon" type="image/png" sizes="256x256" href="//favicon-256-256.png?h=test">",
  "<link rel="icon" type="image/png" sizes="36x36" href="//android-chrome-36-36.png?h=test">",
  "<link rel="icon" type="image/png" sizes="48x48" href="//android-chrome-48-48.png?h=test">",
  "<link rel="icon" type="image/png" sizes="72x72" href="//android-chrome-72-72.png?h=test">",
  "<link rel="icon" type="image/png" sizes="96x96" href="//android-chrome-96-96.png?h=test">",
  "<link rel="icon" type="image/png" sizes="144x144" href="//android-chrome-144-144.png?h=test">",
  "<link rel="icon" type="image/png" sizes="192x192" href="//android-chrome-192-192.png?h=test">",
  "<link rel="icon" type="image/png" sizes="256x256" href="//android-chrome-256-256.png?h=test">",
  "<link rel="icon" type="image/png" sizes="384x384" href="//android-chrome-384-384.png?h=test">",
  "<link rel="icon" type="image/png" sizes="512x512" href="//android-chrome-512-512.png?h=test">",
  "<link rel="apple-touch-icon" sizes="57x57" href="//apple-touch-icon-57-57.png?h=test">",
  "<link rel="apple-touch-icon" sizes="60x60" href="//apple-touch-icon-60-60.png?h=test">",
  "<link rel="apple-touch-icon" sizes="72x72" href="//apple-touch-icon-72-72.png?h=test">",
  "<link rel="apple-touch-icon" sizes="76x76" href="//apple-touch-icon-76-76.png?h=test">",
  "<link rel="apple-touch-icon" sizes="114x114" href="//apple-touch-icon-114-114.png?h=test">",
  "<link rel="apple-touch-icon" sizes="120x120" href="//apple-touch-icon-120-120.png?h=test">",
  "<link rel="apple-touch-icon" sizes="144x144" href="//apple-touch-icon-144-144.png?h=test">",
  "<link rel="apple-touch-icon" sizes="152x152" href="//apple-touch-icon-152-152.png?h=test">",
  "<link rel="apple-touch-icon" sizes="167x167" href="//apple-touch-icon-167-167.png?h=test">",
  "<link rel="apple-touch-icon" sizes="180x180" href="//apple-touch-icon-180-180.png?h=test">",
  "<link rel="apple-touch-icon" sizes="1024x1024" href="//apple-touch-icon.png.png?h=test">",
  "<link rel="apple-touch-icon" sizes="180x180" href="//apple-touch-icon-precomposed.png.png?h=test">",
  "<link rel="apple-touch-icon" media="(device-width: 320) and (device-height: 568) and (orientation: portrait) and (-webkit-device-pixel-ratio: 2)" href="//apple-launch-se-640-1136.png?h=test">",
  "<link rel="apple-touch-icon" media="(device-width: 375) and (device-height: 667) and (orientation: portrait) and (-webkit-device-pixel-ratio: 2)" href="//apple-launch-8-750-1334.png?h=test">",
  "<link rel="apple-touch-icon" media="(device-width: 375) and (device-height: 812) and (orientation: portrait) and (-webkit-device-pixel-ratio: 3)" href="//apple-launch-xs-1125-2436.png?h=test">",
  "<link rel="apple-touch-icon" media="(device-width: 390) and (device-height: 844) and (orientation: portrait) and (-webkit-device-pixel-ratio: 3)" href="//apple-launch-12-1170-2532.png?h=test">",
  "<link rel="apple-touch-icon" media="(device-width: 414) and (device-height: 896) and (orientation: portrait) and (-webkit-device-pixel-ratio: 2)" href="//apple-launch-11-828-1792.png?h=test">",
  "<link rel="apple-touch-icon" media="(device-width: 414) and (device-height: 896) and (orientation: portrait) and (-webkit-device-pixel-ratio: 3)" href="//apple-launch-11-pro-max-1242-2688.png?h=test">",
  "<link rel="apple-touch-icon" media="(device-width: 414) and (device-height: 736) and (orientation: portrait) and (-webkit-device-pixel-ratio: 3)" href="//apple-launch-8-plus-1242-2208.png?h=test">",
  "<link rel="apple-touch-icon" media="(device-width: 414) and (device-height: 736) and (orientation: portrait) and (-webkit-device-pixel-ratio: 3)" href="//apple-launch-6-plus-1242-2208.png?h=test">",
  "<link rel="apple-touch-icon" media="(device-width: 428) and (device-height: 926) and (orientation: portrait) and (-webkit-device-pixel-ratio: 3)" href="//apple-launch-12-pro-max-1284-2778.png?h=test">",
  "<link rel="apple-touch-icon" media="(device-width: 376) and (device-height: 667) and (orientation: portrait) and (-webkit-device-pixel-ratio: 3)" href="//apple-launch-screen-default.png?h=test">",
  "<link rel="apple-touch-icon" href="//apple-launch-screen-default.png?h=test">",
  "<link rel="apple-touch-icon" media="(device-width: 768) and (device-height: 1024) and (orientation: portrait) and (-webkit-device-pixel-ratio: 2)" href="//apple-launch-ipad-pro-s-1536-2048.png?h=test">",
  "<link rel="apple-touch-icon" media="(device-width: 810) and (device-height: 1080) and (orientation: portrait) and (-webkit-device-pixel-ratio: 2)" href="//apple-launch-ipad-pro-1620-2160.png?h=test">",
  "<link rel="apple-touch-icon" media="(device-width: 834) and (device-height: 1194) and (orientation: portrait) and (-webkit-device-pixel-ratio: 2)" href="//apple-launch-ipad-pro-l-1668-2388.png?h=test">",
  "<link rel="apple-touch-icon" media="(device-width: 834) and (device-height: 1112) and (orientation: portrait) and (-webkit-device-pixel-ratio: 2)" href="//apple-launch-air-1668-2224.png?h=test">",
  "<link rel="apple-touch-icon" media="(device-width: 1024) and (device-height: 1366) and (orientation: portrait) and (-webkit-device-pixel-ratio: 2)" href="//apple-launch-ipad-pro-xl-2048-2732.png?h=test">",
  "<link rel="shortcut icon" href="//favicon.ico?h=test">",
]
`;
